### PR TITLE
Add well-formedness test for assume/assert

### PR DESCRIPTION
**Pull request by xbauch**
Two simple tests are added: 1) targets are empty, 2) the guard evaluates to a
Boolean. The second test is a case-analysis of the expression id.

### CODE REVIEW

File: src/goto-programs/goto_program.cpp [hunk 1]
```diff
@@ -668,6 +668,91 @@ bool goto_programt::instructiont::equals(const instructiont &other) const
...
+    break;
+  case ASSUME:
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      targets.empty(),
+      "assume instruction should not have a target",
+      source_location);
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      evaluates_to_boolean(guard),
+      "assuming non-boolean condition\n" + guard.pretty(),
+      source_location);
```
Comment (danpoe) [inline src/goto-programs/goto_program.cpp:739]: Btw, if the instruction that we're checking has a corresponding code class in `std_code.h` that is used in the `code` field of the instruction, we should also add `check()/validate()/validate_full()` methods to that class.

For example, an assignment instruction contains a `code_assignt` object in its `code` field. However, many instructions don't actually fill in the `code` field I think. For example, I think an assume instruction does not contain a `code_assumet` object in its `code` field (as the `code_assumet` class is mostly only used to represent the result of parsing C code, but for the actual instruction it is not needed as it already has a `guard` which can hold the condition of the `assume`). In such cases we do not need to add the `check()/validate()/validate_full()` methods to `code_assumet`.

File: src/goto-programs/goto_program.cpp [hunk 2]
```diff
@@ -755,13 +755,29 @@ void goto_programt::instructiont::validate(
...
       evaluates_to_boolean(guard),
       "goto with non-boolean condition\n" + guard.pretty(),
       source_location);
     break;
+  case FUNCTION_CALL:
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      code.get_statement() == ID_function_call,
+      "function call instruction should contain a call statement",
+      source_location);
+    code.validate(ns, vm);
```
Comment (danpoe) [inline src/goto-programs/goto_program.cpp]: The method `validate_full()` of `code` is already called by the call to `validate_code_full()` above so I think we can omit this one.

File: src/goto-programs/goto_program.cpp [hunk 3]
```diff
@@ -755,13 +755,29 @@ void goto_programt::instructiont::validate(
...
+    break;
+  case RETURN:
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      code.get_statement() == ID_return,
+      "return instruction should contain a return statement",
+      source_location);
+    code.validate(ns, vm);
```
Comment (danpoe) [inline src/goto-programs/goto_program.cpp]: Can be omitted.

File: src/util/std_code.h [hunk 4]
```diff
@@ -1052,6 +1052,40 @@ class code_function_callt:public codet
   {
     return op2().operands();
   }
+
+  void check(const validation_modet vm = validation_modet::INVARIANT) const
+  {
+    DATA_CHECK(
+      operands().size() == 3, "function calls must have three operand");
```
Comment (danpoe) [inline src/util/std_code.h]: `operand` -> `operands`

File: src/util/std_code.h [hunk 5]
```diff
@@ -1052,6 +1052,40 @@ class code_function_callt:public codet
...
+  }
+
+  void validate(
+    const namespacet &ns,
+    const validation_modet vm = validation_modet::INVARIANT) const
+  {
+    check(vm);
+    if(lhs().id() != ID_nil)
+      DATA_CHECK(
+        base_type_eq(lhs().type(), function().type(), ns),
```
Comment (danpoe) [inline src/util/std_code.h]: I think we need to use `to_code_type(function().type()).return_type()` here.

Also we could additionally check that if `lhs.id() == ID_nil` then the return type of the call is `ID_empty` (which represents `void`).

File: src/util/std_code.h [hunk 6]
```diff
@@ -1052,6 +1052,40 @@ class code_function_callt:public codet
...
+  {
+    for(const exprt &op : operands())
+    {
+      validate_expr_full_pick(op, ns, vm);
+    }
+
+    for(const exprt &arg : op2().operands())
+    {
+      validate_expr_full_pick(arg, ns, vm);
+    }
```
Comment (danpoe) [inline src/util/std_code.h]: I think we can omit this loop as the iteration over `op2().operands()` would already be done by the loop above which recursively calls `validate_full()` for all the operands of the expression.

File: unit/goto-programs/goto_program_function_call.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+#include <util/arith_tools.h>
+
+SCENARIO(
+  "Validation of well-formed function call codes",
+  "[core][goto-programs][validate]")
+{
+  GIVEN("A program with one function call")
+  {
+    symbol_tablet symbol_table;
+    const typet type1 = signedbv_typet(32);
```
Comment (danpoe) [inline unit/goto-programs/goto_program_function_call.cpp]: I'd prefer using `const signedbv_typet type1` here. (though ultimately we'll have to assign this type to a `typet` anyways when assigining to the `type` field of a `symbolt`).

**Review by danpoe — COMMENTED** [review]
Looks good to me overall. A few minor comments.

File: src/util/std_code.h (hunk 5 above)
Comment (xbauch) [reply src/util/std_code.h]: I actually managed to trigger this type inconsistency in my unit test but I will try with the case as well.

**Pull request closed without merging**